tests.nixpkgs-check-by-name: Syntactic `callPackage` detection and allow new package variants

infinisil opened the pull request:
## Description of changes

This fixes a false positive with the `pkgs/by-name` CI check, where it thinks that a new package defined using `callPackage` is introduced, but it's really just something that happens to use `callPackage` underneath.

This caused the check to mistakenly fail for at least these PRs:
- https://github.com/NixOS/nixpkgs/pull/256115
- https://github.com/NixOS/nixpkgs/pull/278607
- https://github.com/NixOS/nixpkgs/pull/282929
- https://github.com/NixOS/nixpkgs/pull/284591
- https://github.com/NixOS/nixpkgs/pull/284602
- https://github.com/NixOS/nixpkgs/pull/286267
- https://github.com/NixOS/nixpkgs/pull/285681

The fix is very involved, because the new code parses the file where each attribute is defined (usually `all-packages.nix`) and syntactically checks whether its definition location is of the form `<attr> = callPackage <file> <arg>`.

The nice thing about this is that this pretty much this code has to be written anyways in order to do the [automated migration](https://github.com/NixOS/nixpkgs/pull/211832).

Furthermore, the last commit fixes another small problem that causes the check to fail for https://github.com/NixOS/nixpkgs/pull/284964, since the same code is touched for these fixes.

Best reviewed commit-by-commit

This work is sponsored by [Antithesis](https://antithesis.com/) and [Tweag](https://www.tweag.io/) :sparkles:

## Things done

- Added tests for the changed/added code
- Added comments and documentation to the changed/added code

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
tests.nixpkgs-check-by-name: LineIndex functionality, semantics, tests

- `fromlinecolumn` is added to be almost the reverse of `line`.
  This is needed in the future to get from `builtins.unsafeGetAttrPos`
  locations to the index for rnix
- Previously the semantics for newline indices were not really defined,
  now they are, and make more sense.
- Now there's a unit test for these functions

infinisil committed:
tests.nixpkgs-check-by-name: Syntactic callPackage detection

This makes the callPackage detection stronger by syntactically detecting
whether an attribute is using callPackage

See the added test case for why this is needed.

infinisil committed:
tests.nixpkgs-check-by-name: Re-use nixFileCache more

Makes the reference check use the nixFileCache instead of separately
parsing and resolving paths

infinisil committed:
tests.nixpkgs-check-by-name: Format

With `cargo fmt`

Explicitly didn't do that for previous commits in order to keep the diff
more easily reviewable

infinisil committed:
tests.nixpkgs-check-by-name: Allow new package variants

adamcstephens reviewed (approved):
I’m approving even though I will admit I haven’t read all the code. Broken checks are bad and I trust your diligence and responsiveness.

philiptaron reviewed (changes requested):
Whew! Looks mostly great. I think I want:

1. Not to throw away the parse error.
2. Move `indoc` to dev-dependencies.
3. Clarify a couple bail-vs. Ok(None) choices

... and I'm super sold.

Thanks for this work.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/utils.rs:
```diff
@@ -35,17 +35,61 @@ impl LineIndex {
         for split in s.split_inclusive('\n') {
             index += split.len();
-            newlines.push(index);
+            newlines.push(index - 1);
```
Kind of. Only the behavior for indices of newline characters themselves is changed. Previously they were thought to be on the succeeding line, but I thought it makes more sense to have them on the preceding line. I don't think this is even used anywhere, but it made more sense when writing the tests and was previously unspecified :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -184,17 +197,56 @@ pub fn check_values(
+                                    location.column,
+                                    nixpkgs_path,
+                                )? {
+                                    None => Success(NonApplicable),
```
Adding a comment, this happens when the definition is not of the form `<attr> = callPackage <arg1> <arg2>;`

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -280,9 +335,10 @@ pub fn check_values(
-            check_result.map(|value| (attribute_name.clone(), value))
-        },
-    ));
+            Ok::<_, anyhow::Error>(check_result.map(|value| (attribute_name.clone(), value)))
```
Ah it works if I remove it now! The type inference struggled, I guess before I actually added any `Err` case :)

infinisil replied:
Oh actually no this problem came back after I removed the only explicit `anyhow::bail` (which converted the `NixFileResult::Err` into an `anyhow::Result`, see https://github.com/NixOS/nixpkgs/pull/285089#discussion_r1477470990):

```
error[E0283]: type annotations needed
   --> src/eval.rs:332:13
    |
332 |             Ok(check_result.map(|value| (attribute_name.clone(), value)))
    |             ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
333 |         })
334 |         .collect_vec()?,
    |                       - type must be known at this point
    |
    = note: multiple `impl`s satisfying `anyhow::Error: From<_>` found in the following crates: `anyhow`, `core`:
            - impl<E> From<E> for anyhow::Error
              where E: StdError, E: Send, E: Sync, E: 'static;
            - impl<T> From<!> for T;
            - impl<T> From<T> for T;
    = note: required for `std::result::Result<Validation<Nixpkgs>, anyhow::Error>` to implement `FromResidual<std::result::Result<Infallible, _>>`
help: consider specifying the generic arguments
    |
332 |             Ok::<Validation<(std::string::String, Package)>, E>(check_result.map(|value| (attribute_name.clone(), value)))
    |               +++++++++++++++++++++++++++++++++++++++++++++++++

```

I'll keep the fish for now :laughing:

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+}
+
+/// A type to represent the result when trying to initialise a `NixFile`.
+type NixFileResult = Result<NixFile, rnix::parser::ParseError>;
```
Originally I was thinking that consumers might want to be able to distinguish the parse error and handle it in different ways, but that's really not necessary after all. I'll fold this into `anyhow::Result` :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+        column: usize,
+        relative_to: &Path,
+    ) -> anyhow::Result<Option<CallPackageArgumentInfo>> {
+        let Some(attrpath_value) = self.attrpath_value_at(line, column)? else {
+            return Ok(None);
+        };
+        self.attrpath_value_call_package_argument_info(attrpath_value, relative_to)
```
Doesn't quite work here unfortunately because `attrpath_value_call_package_argument_info` returns a `Result<Option<..>, ..>`

I could use a `match` or `if let Some` instead here, but I felt like using the same style as in the code below

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+            )
+        };
+
+        if attrpath_node.kind() != SyntaxKind::NODE_ATTRPATH {
+            return Ok(None);
+        }
```
Adding a comment here, this is to handle `inherit`'s

infinisil replied:
The distinction regarding `None` vs `Err` is that:
- `None` is explained in the method doc comment:
  > If there's no guaranteed `callPackage` at that location, `None` is returned.
  
  This means it's something that the PR author can influence, and this might be handled in different ways (in this case, nothing is done for such cases).

- `Err` is for programmatic errors, indicating a bug that's not caused by the PR author. For example I/O errors, rnix parse errors (since cpp Nix will have already parsed it before, it means rnix has a bug), or often the case in this function, `rnix` AST fields are empty when they really shouldn't be (no idea why rnix has all AST fields optional..)

infinisil replied:
I'm adding a comment on this

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+            )
+        };
+
+        if !ast::AttrpathValue::can_cast(attrpath_value_node.kind()) {
+            anyhow::bail!(
+                "Node in {} is not an attribute path value node: {attrpath_value_node:?}",
+                self.path.display()
+            )
+        }
```
Confusingly, there's both `NODE_ATTRPATH` (e.g. `foo.bar`) and `NODE_ATTRPATH_VALUE` (`foo.bar = some_value;`). Here we're validating that the parent of the `NODE_ATTRPATH` is a `NODE_ATTRPATH_VALUE` :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+        }
+        // attrpath_value_node looks like "foo.bar = 10;"
+
+        // unwrap is fine because we confirmed that we can cast with the above check
+        Ok(Some(ast::AttrpathValue::cast(attrpath_value_node).unwrap()))
```
Unfortunately `cast` consumes `attrpath_value_node`, so if we try to do

```rust
        if let Some(attrpath_value_ast_node) = ast::AttrpathValue::cast(attrpath_value_node) {
            Ok(Some(attrpath_value_ast_node))
        } else {
            anyhow::bail!(
                "Node in {} is not an attribute path value node: {attrpath_value_node:?}",
                self.path.display()
            )
        }
```

We'll get a borrow checker error:

```
    Checking nixpkgs-check-by-name v0.1.0 (/home/tweagysil/src/nixpkgs/by-name/pkgs/test/nixpkgs-check-by-name)
error[E0382]: borrow of moved value: `attrpath_value_node`
   --> src/nix_file.rs:180:66
    |
169 |         let Some(attrpath_value_node) = attrpath_node.parent() else {
    |                  ------------------- move occurs because `attrpath_value_node` has type `rowan::SyntaxNode<NixLanguage>`, which does not implement the `Copy` trait
...
176 |         if let Some(attrpath_value_ast_node) = ast::AttrpathValue::cast(attrpath_value_node) {
    |                                                                         ------------------- value moved here
...
180 |                 "Node in {} is not an attribute path value node: {attrpath_value_node:?}",
    |                                                                  ^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after move
    |
    = note: this error originates in the macro `$crate::__export::format_args` which comes from the expansion of the macro `anyhow::bail` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value if the performance cost is acceptable
    |
176 |         if let Some(attrpath_value_ast_node) = ast::AttrpathValue::cast(attrpath_value_node.clone()) {
    |                                                                                            ++++++++
```


Could do the `clone` as suggested by the error message, but feels a bit pointless to clone when it shouldn't be necessary. I believe the `rnix` crate should really have a way to work with non-owned values.

infinisil replied:
For now I'm adding a comment explaining this here.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+            if let ResolvedPath::Within(p) = self.static_resolve_path(actual_path, relative_to) {
+                Some(p)
+            } else {
+                // We can't statically know an existing path inside Nixpkgs used as <arg2>
+                None
```
The one legitimate case might be `./foo/${bar}/baz`, which we can't resolve statically, but could be a totally valid path. Cases like search paths (`<nixpkgs>`) could also occur theoretically, but I hope nobody uses that. Same for cases of paths outside Nixpkgs, pointing into `/nix/store` for example.

These are fairly odd cases, but I think the code here should not be concerned about handling these. It really just needs to check if the attribute is of the form `<attr> = callPackage <arg1> <arg2>` and return information about `<arg1>` and `<arg2>`.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+                None
+            }
+        } else {
+            // <arg2> is not a path, but rather e.g. an inline expression
+            None
```
See https://github.com/NixOS/nixpkgs/pull/285089#discussion_r1477503041

This case does indeed occur though, e.g. https://github.com/NixOS/nixpkgs/blob/7ece42702159f2cca6f7f2238d32be4f1830eee1/pkgs/top-level/all-packages.nix#L171-L177

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+                // callPackage-mocking method will take care of not triggering for this case.
+
+                if select.default_expr().is_some() {
+                    // Very odd case, but this would be `foo = self.callPackage or true ./test.nix {}
+                    // (yes this is valid Nix code)
+                    return Ok(None);
```
In spirit of https://github.com/NixOS/nixpkgs/pull/285089#discussion_r1477482119 I'd like to keep the `None` here.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+                if let ast::Attr::Ident(ident) = last {
+                    ident
+                } else {
+                    // Here it's something like `foo = self."callPackage" /test.nix {}`
+                    // which we're not gonna bother with
+                    return Ok(None);
```
Also see https://github.com/NixOS/nixpkgs/pull/285089#discussion_r1477482119

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+                    return Ok(None);
+                }
+            }
+            // Any other expression we're not gonna treat as callPackage
+            _ => return Ok(None),
```
Yeah this is pretty big, it's all the different AST variants: https://docs.rs/rnix/latest/rnix/ast/enum.Expr.html

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+    Interpolated,
+    /// Something like `<nixpkgs>`, can't be known statically
+    SearchPath,
+    /// Path couldn't be resolved due to an IO error,
+    /// e.g. if the path doesn't exist or you don't have the right permissions
+    Unresolvable(std::io::Error),
```
I have seen it occur in PRs before, e.g. when you forget to `git add` files :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+    ///
+    /// E.g. for the path expression `./bar.nix` in `./foo.nix` and an absolute path of the
+    /// current directory, the function returns `ResolvedPath::Within(./bar.nix)`
+    pub fn static_resolve_path(&self, node: ast::Path, relative_to: &Path) -> ResolvedPath {
```
Hmm I'm not sure, it definitely can't resolve arbitrary Nix path expressions, because that would require Nix evaluation. So static here refers to not doing any evaluation.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/nix_file.rs:
```diff
@@ -0,0 +1,508 @@
+        }
+
+        // Join the file's parent directory and the path expression, then resolve it
+        // FIXME: Expressions like `../../../../foo/bar/baz/qux` or absolute paths
+        // may resolve close to the original file, but may have left the relative_to.
+        // That should be checked more strictly
```
Indeed. I have some [older code](https://github.com/nixpkgs-architecture/nix-spp/blob/2a6ff6cb2a74f55032aa48531eac5a14dc4fc2bb/src/index.rs#L199) that implements this, but that's for another PR.

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/references.rs:
```diff
@@ -1,23 +1,29 @@
+        nix_file_cache,
+        relative_package_dir,
+        absolute_package_dir,
+        Path::new(""),
```
Expanding this, agree that this is a bit special and deserves a better comment :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/references.rs:
```diff
@@ -63,20 +70,27 @@ fn check_path(
+                        absolute_package_dir,
+                        &entry_subpath,
+                    )
+                    .with_context(|| format!("Error while recursing into {}", subpath.display()))
```
Ah good point. it's the context for the recursion into the directory, so it should be for `subpath`. I'm moving the `with_context` one level up to make this more obvious.

infinisil committed:
tests.nixpkgs-check-by-name: Apply feedback from review

https://github.com/NixOS/nixpkgs/pull/285089#pullrequestreview-1861099233

Many thanks, Philip Taron!

infinisil committed:
tests.nixpkgs-check-by-name: Refactor eval.rs

There was too much indentation!

infinisil commented:
@philiptaron Thanks a lot! Addressed all your comments in two new commits :)

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -184,17 +197,56 @@ pub fn check_values(
+                                        }
+                                    },
+                                },
+                        },
```
Now refactored this function by adding two new functions to handle the by-name and non-by-name cases separately :)

philiptaron reviewed (approved):
Ship it.

tomberek commented:
My understanding was that this check would only apply if the "extra" parameter was empty `{}`. Thus exempting cases like:
```
new-package = callPackage ./pkgs/somewhere.nix { some-arg = true;};
```

infinisil commented:
@tomberek It applies to all new packages using `callPackage` (even before this PR, this is just a bug fix). It's not a problem to require this for a non-empty second argument because we support overriding it for packages in `pkgs/by-name`, see [here](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#changing-implicit-attribute-defaults).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-54/39990/1
